src/NHibernate/Dialect/SQLiteDialect.cs

```diff
@@ -94,8 +102,34 @@ protected virtual void RegisterFunctions()
 
 			// NH-3787: SQLite requires the cast in SQL too for not defaulting to string.
 			RegisterFunction("transparentcast", new CastFunction());
-			
-			RegisterFunction("strguid", new SQLFunctionTemplate(NHibernateUtil.String, "substr(hex(?1), 7, 2) || substr(hex(?1), 5, 2) || substr(hex(?1), 3, 2) || substr(hex(?1), 1, 2) || '-' || substr(hex(?1), 11, 2) || substr(hex(?1), 9, 2) || '-' || substr(hex(?1), 15, 2) || substr(hex(?1), 13, 2) || '-' || substr(hex(?1), 17, 4) || '-' || substr(hex(?1), 21) "));
+
+			if (_binaryGuid)
+				RegisterFunction("strguid", new SQLFunctionTemplate(NHibernateUtil.String, "substr(hex(?1), 7, 2) || substr(hex(?1), 5, 2) || substr(hex(?1), 3, 2) || substr(hex(?1), 1, 2) || '-' || substr(hex(?1), 11, 2) || substr(hex(?1), 9, 2) || '-' || substr(hex(?1), 15, 2) || substr(hex(?1), 13, 2) || '-' || substr(hex(?1), 17, 4) || '-' || substr(hex(?1), 21) "));
+			else
+				RegisterFunction("strguid", new SQLFunctionTemplate(NHibernateUtil.String, "cast(?1 as char)"));
+		}
+
+
+		public override void Configure(IDictionary<string, string> settings)
+		{
+			base.Configure(settings);
+
+			if (settings.TryGetValue(Cfg.Environment.ConnectionString, out string connectionString))
```

Comment:
Shouldn't we also check for `Cfg.Environment.ConnectionStringName` as it's done in [`ConnectionProvider`](https://github.com/nhibernate/nhibernate-core/blob/27ea900b6ebaf5de5e9ca603ae57e6ca5197f968/src/NHibernate/Connection/ConnectionProvider.cs#L51-L62)?

I also wonder how to make it work with [multi-tenancy](https://github.com/nhibernate/nhibernate-core/pull/2108) where connection is provided externally... It seems default connection must be provided with proper configuration.

Reply:
As for the first comment, it seems we should extract some code from ConnectionProvider to have something like Environment.ResolveConnectionString(...)

Reply:
Maybe... Or just do copy/paste as you won't be able to reuse this logic in `ConnectionProvider` as `GetNamedConnectionString` can be overridden.

Reply:
It's a mess.

On a tangential note, Configuration.BuildSessionFactory() calls BuildSettings() which calls GetDerivedProperties(), which creates a temporary instance of the dialect to access any default properties set up by it, and combines those with the actual configuration. Then the whole thing gets sent to SettingsFactory.BuildSettings(), which, as the very first order of business, creates the real instance of the dialect _including_ reading its default properties and combining them with the actual configuration once more.

Anyway, back to ConnectionProvider. No idea why GetNamedConnectionString() is virtual. It is not overridden within NHibernate. You also can override both the top-level (public) Configure() and GetConnection(). One would think those options would be enough. I suspect someone just slapped virtual on everything. Would probably make more sense to combine reading both settings into a GetConnectionString() that might be virtual.

ConnectionProvider does only read the connection string once during Configure() and stores the effective value in its property `ConnectionString`. This is not available to the dialect because it doesn't have the ConnectionProvider and also the dialect is created earlier than the connection provider.

So, teaching the dialect about named connection strings is simpler but of course breaks down if anyone does in fact override GetNamedConnectionString() to get the configuration from somewhere else. And as @bahusoid says, it's still not enough for multitenancy.

I'll look into adding a dialect-specific setting to be used for scenarios where the connection string isn't enough.